- Scikit learn is the machine learning library of Python 
- There are also a few (7) in built datasets which are very helpful to begin with understanding the concepts and also to build a quick POC.

In [3]:
import sklearn
print(sklearn.__version__)

1.6.1


**how to load a csv file that has a space delimiter and that exists without column names?**

In [10]:
import pandas as pd
boston_csv_df2 = pd.read_csv('HousingData.csv')

In [11]:
boston_csv_df2.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2

In [12]:
boston_csv_df2.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT', 'MEDV'],
      dtype='object')

**how to load a csv file?**

In [13]:
boston_csv_df2.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS       float64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD          int64
TAX          int64
PTRATIO    float64
B          float64
LSTAT      float64
MEDV       float64
dtype: object

In [14]:
boston_csv_df2.shape

(506, 14)

In [17]:
boston_csv_df2.isnull().sum()

CRIM        0
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

In [40]:
for column in boston_csv_df2.columns:
    if boston_csv_df2[column].isna().sum() > 0:
        median_value = boston_csv_df2[column].median()
        boston_csv_df2[column].fillna(median_value, inplace=True)
print(" Missing values after imputation:")
print(boston_csv_df2.isna().sum())

 Missing values after imputation:
CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64


In [41]:
boston_csv_df2.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

From the dataset, we conclude the following:
- There are 506 rows and 14 columns in total.
- MEDV column represents the target variable.

# Split data into X & y

In [42]:
X = boston_csv_df2.drop("MEDV", axis=1) #mapping only the features to the X variable
y = boston_csv_df2["MEDV"] #mapping the target column to y variable

In [43]:
X.shape

(506, 13)

In [44]:
y.shape

(506,)

# Split data into train and test set

In [45]:
from sklearn.model_selection import train_test_split

In [46]:
X_train, X_test, y_train,y_test = train_test_split(X,y,test_size=0.2, random_state=21)

test_size is a parameter tht can accept any value between 0 and 1. But mind that 0 and 1 are exclude (i.e) it can accept any value>0 and <1

for ex, if test_size = 0.2, it implies that 20% of the data points are to be used for testing and the remaining 80% of the data points are to be used for training.

for ex, if test_size = 0.1, it implies that 10% of the data points are to be used for testing and the remaining 90% of the data points are to be used for training.

- test_size = 0 ==> None of the data points to be used for testing
- test_size = 1 ==> 100% of data to be used for testing  ==> this is the reason for the test_size parameter to never accept the values 0 and 1.


In [47]:
print("Shape of train set features")
print(X_train.shape)
print("Shape of test set features")
print(X_test.shape)

Shape of train set features
(404, 13)
Shape of test set features
(102, 13)


In [48]:
print("Shape of train set target")
print(y_train.shape)
print("Shape of test set target")
print(y_test.shape)

Shape of train set target
(404,)
Shape of test set target
(102,)


# Building the Multiple Linear Regression model
- Step 1 : Import a model of our choice.
- Step 2 : Instantiate an estimator object which contains the ML algorithm.
- Step 3 : Fit the ML algorithm on the given data (train set)
- Step 4 : Use the generated model on the test set for predictions.
- Step 5 : Evaluate your model using appropriate evaluation metrics.

In [49]:
from sklearn.linear_model import LinearRegression

In [50]:
model_lr = LinearRegression()

In [51]:
model_lr.fit(X_train,y_train) #supervised machine learning, checks how is the relationship between the features(X_train) and the output(y_train)

LinearRegression()

In [52]:
print("The intercept for this Multiple Linear regression model on the boston dataset is ", 
      model_lr.intercept_)

The intercept for this Multiple Linear regression model on the boston dataset is  39.35952555211737


In [53]:
print("The regression co-efficients of the Multiple Linear regression model are :")
print("==========================================================================")
print(model_lr.coef_)

The regression co-efficients of the Multiple Linear regression model are :
[-9.38219654e-02  4.63139008e-02 -3.68402826e-02  3.51633568e+00
 -1.71515915e+01  3.52655053e+00 -9.62281251e-03 -1.49385068e+00
  2.86917200e-01 -1.02083215e-02 -1.03180158e+00  1.04996827e-02
 -4.97221467e-01]


In [54]:
y_test_predicted_values = model_lr.predict(X_test)
#Always only the test data features should be passed as input to the predict method

In [55]:
y_test_predicted_values

array([15.3223    , 16.00629672, 26.98055685, 37.37461827, 33.46090579,
       12.5881884 , 25.79888207, 30.24493819, 33.50522764, 18.35693554,
       23.08221157, 16.42984613, 28.75742883,  9.42354175, 18.59396763,
       28.04272061, 15.84856407, 29.12351887,  4.15529599, 17.82550341,
       20.96264152, 26.513564  , 19.32455247,  2.1832943 , 20.797066  ,
       34.29891841, 25.36157885, 31.55745667, 12.79390151, 25.11239047,
       25.44857706, 18.73572605, 34.17861271, 13.3417708 , 24.82788251,
       31.42534765, 31.84685041, 21.76721037, 21.01457388, 17.48785193,
       19.63958819, 32.57665625, 24.48559205, 24.56698283, 22.16280824,
       20.61021339, 30.27309849, 21.32013646, 30.72014245, 23.99217324,
       34.50604195, 14.68982621, 41.18335933, 16.4384516 , 13.70225534,
       11.41734289,  8.24392723, 23.51022683, 22.60199368, 27.18606401,
       17.01505605, 35.25768973, 25.15612307, 22.66796622, 35.64950567,
       36.54506275, 27.82215999, 19.27858487, 21.52332253, 19.53

In [56]:
y_test_predicted_values.shape

(102,)

In [57]:
import pandas as pd
pd.DataFrame({'Actual y_test': y_test, 'Predicted y_test':y_test_predicted_values})

Actual y_test  Predicted y_test
455           14.1         15.322300
142           13.4         16.006297
311           22.1         26.980557
232           41.7         37.374618
290           28.5         33.460906
..             ...               ...
486           19.1         19.575229
468           19.1         17.255761
302           26.4         29.223185
244           17.6         16.483318
321           23.1         24.827860

[102 rows x 2 columns]

# Evaluating the MLR model

In [58]:
print("R2 score is:{}".format(model_lr.score(X_test,y_test)))

R2 score is:0.6884038306845459


In [59]:
from sklearn.metrics import mean_squared_error
import numpy as np

In [60]:
print("RMSE value of the test dataset is :")
print(np.sqrt(mean_squared_error(y_test, y_test_predicted_values))) #(actual values,predicted values)

RMSE value of the test dataset is :
5.414997899067064


# End of discussion